psi4/driver/qcdb/molecule.py

```diff
@@ -1336,6 +1336,96 @@ def run_dftd3(self, func: str = None, dashlvl: str = None, dashparam: Dict = Non
         elif derint == 1:
             return jobrec['extras']['qcvars']['DISPERSION CORRECTION GRADIENT']
 
+    def run_dftd4(self, func=None, dashlvl=None, dashparam=None, dertype=None, verbose=1):
+        """Compute dispersion correction via Grimme's DFTD4 program.
+
+        Parameters
+        ----------
+        func : str, optional
+            Name of functional (func only, func & disp, or disp only) for
+            which to compute dispersion (e.g., blyp, BLYP-D2, blyp-d3bj,
+            blyp-d3(bj), hf+d). Any or all parameters initialized
+            from `dashcoeff[dashlvl][func]` can be overwritten via
+            `dashparam`.
+        dashlvl : str, optional
+            Name of dispersion correction to be applied (e.g., d, D2,
+            d3(bj), das2010). Must be key in `dashcoeff` or "alias" or
+            "formal" to run.
+        dashparam : dict, optional
+            Values for the same keys as `dashcoeff[dashlvl]['default']`
+            used to override any or all values initialized by `func`.
+            Extra parameters will error.
+        dertype : int or str, optional
+            Maximum derivative level at which to run DFTD3. For large
+            molecules, energy-only calculations can be significantly more
+            efficient. Influences return values, see below.
+        verbose : int, optional
+            Amount of printing.
+
+        Returns
+        -------
+        energy : float
+            When `dertype=0`, energy [Eh].
+        gradient : ndarray
+            When `dertype=1`, (nat, 3) gradient [Eh/a0].
+        (energy, gradient) : tuple of float and ndarray
+            When `dertype=None`, both energy [Eh] and (nat, 3) gradient [Eh/a0].
+
+        Notes
+        -----
+        This function wraps the QCEngine dftd4 harness which wraps the internal DFTD4 Python API.
+        As such, the upstream convention of `func` trumping `dashparam` holds, rather than the
+        :py:func:`run_dftd3` behavior of `dashparam` extending or overriding `func`.
```

Comment:
Is there any way to harmonise these two? The way I understand it, we do the same thing Psi4-side, but the behaviour of dftd3 and dftd4 differs when both func and dashparams are provided.

In any case, not Psi4's but rather qcng's problem...

I'm guessing the only problem might be when one wants to run, say, B3LYP (for which `func` values are defined) with a custom set of parameters - is this possible to do with dftd4?

Reply:
See related discussion here: https://github.com/dftd4/dftd4/issues/98

Short summary:

`dftd4` does not provide the possibility to select a functional, say B3LYP and only change the *s8*, this is by design. Either the downstream user hands the responsibility over to `dftd4` to select the correct parameters or provides a complete damping parameters.

If Psi4 requires full control over the parameter selection, it can't use the former mechanism, but has to always construct the set of damping parameters first (`qcng` knows how to do this for all supported functionals in `dftd4`).

Note that in the future `dftd4` will most likely throw an error in case both functional name and an (incomplete) set of damping parameters are provided.